server/src/main/java/com/google/musicanalysis/site/OAuthCallbackServlet.java

```diff
@@ -37,40 +37,51 @@ protected void doGet(HttpServletRequest req, HttpServletResponse res)
     }
 
     var sessionOauthState = req.getSession().getAttribute("oauth:state");
+    var sessionOauthService = req.getSession().getAttribute("oauth:service");
 
     if (!state.equals(sessionOauthState)) {
       res.setStatus(401);
       return;
     }
 
+    String clientSecret;
+    String clientId;
+    URI tokenUri;
+
+    if (sessionOauthService.equals("youtube")) {
```

Comment:
Having both the youtube and spotify oauth callbacks together in the same class hinders long-term maintenance since it becomes hard to detangle these two code paths. If you want to add a new spotify feature, you'll have to explicitly hide it from the youtube code path. Consider creating an abstract oauth class and having two separate routes/classes:
```
@WebServlet("/api/oauth/callback-spotify")
public class SpotifyOauthCallbackServlet extends OauthCallbackServlet {
  protected override clientSecret = System.getenv().get("SPOTIFY_CLIENT_SECRET");
  protected override clientId = SPOTIFY_CLIENT_ID;
  protected override tokenUri = URI.create("https://accounts.spotify.com/api/token");
  ...
}
```

This will give you more flexibility to branch more functionality as needed.